File: ALittle/Module/ALittleIDE/Other/GameLibrary/Std/JSScript/Std/Loop/LoopTimer.js
{
if (typeof ALittle === "undefined") window.ALittle = {};


if (ALittle.LoopObject === undefined) throw new Error(" extends class:ALittle.LoopObject is undefined");
ALittle.LoopTimer = JavaScript.Class(ALittle.LoopObject, {
	Ctor : function(func, delay_time) {
		this._func = func;
		if (delay_time === undefined || delay_time < 0) {
			delay_time = 0;
		}
		this._total_delay_time = delay_time;
		this._accumulate_count = 0;
		this._accumulate_delay_time = 0;
		if (this._func === undefined) {
			this._accumulate_count = 1;
			this._accumulate_delay_time = this._total_delay_time;
			ALittle.Log("LoopTimer create failed:function is nil or not a function");
			return;
		}
		this._complete_callback = undefined;
	},
	get complete_callback() {
		return this._complete_callback;
	},
	set complete_callback(value) {
		this._complete_callback = value;
	},
	Reset : function() {
		this._accumulate_count = 0;
		this._accumulate_delay_time = 0;
	},
	Completed : function() {
		if (this._complete_callback !== undefined) {
			this._complete_callback();
		}
	},
	IsCompleted : function() {
		return this._accumulate_count >= 1;
	},
	SetCompleted : function() {
		if (this._accumulate_count >= 1) {
			return;
		}
		this._accumulate_count = 1;
		this._accumulate_delay_time = this._total_delay_time;
		this._func();
	},
	SetTime : function(time) {
		this._accumulate_count = 0;
		this._accumulate_delay_time = 0;
		if (time <= 0) {
			return [0, false];
		}
		if (time <= this._total_delay_time) {
			this._accumulate_delay_time = time;
			return [0, false];
		}
		this._accumulate_delay_time = this._total_delay_time;
		this._accumulate_count = 1;
		this._func();
		return [time - this._total_delay_time, true];
	},
	Update : function(frame_time) {
		if (this._accumulate_delay_time < this._total_delay_time) {
			this._accumulate_delay_time = this._accumulate_delay_time + (frame_time);
			if (this._accumulate_delay_time < this._total_delay_time) {
				return;
			}
			this._accumulate_delay_time = this._total_delay_time;
		}
		if (this._accumulate_count >= 1) {
			return;
		}
		this._accumulate_count = 1;
		this._func();
	},
}, "ALittle.LoopTimer");

}

File: ALittle/Module/Std/Script/Std/Loop/LoopTimer.lua
-- ALittle Generate Lua
module("ALittle", package.seeall)

local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs


assert(ALittle.LoopObject, " extends class:ALittle.LoopObject is nil")
LoopTimer = Lua.Class(ALittle.LoopObject, "ALittle.LoopTimer")

function LoopTimer:Ctor(func, delay_time)
	___rawset(self, "_func", func)
	if delay_time == nil or delay_time < 0 then
		delay_time = 0
	end
	___rawset(self, "_total_delay_time", delay_time)
	___rawset(self, "_accumulate_count", 0)
	___rawset(self, "_accumulate_delay_time", 0)
	if self._func == nil then
		___rawset(self, "_accumulate_count", 1)
		___rawset(self, "_accumulate_delay_time", self._total_delay_time)
		Log("LoopTimer create failed:function is nil or not a function")
		return
	end
	___rawset(self, "_complete_callback", nil)
end

function LoopTimer.__getter:complete_callback()
	return self._complete_callback
end

function LoopTimer.__setter:complete_callback(value)
	self._complete_callback = value
end

function LoopTimer:Reset()
	self._accumulate_count = 0
	self._accumulate_delay_time = 0
end

function LoopTimer:Completed()
	if self._complete_callback ~= nil then
		self._complete_callback()
	end
end

function LoopTimer:IsCompleted()
	return self._accumulate_count >= 1
end

function LoopTimer:SetCompleted()
	if self._accumulate_count >= 1 then
		return
	end
	self._accumulate_count = 1
	self._accumulate_delay_time = self._total_delay_time
	self._func()
end

function LoopTimer:SetTime(time)
	self._accumulate_count = 0
	self._accumulate_delay_time = 0
	if time <= 0 then
		return 0, false
	end
	if time <= self._total_delay_time then
		self._accumulate_delay_time = time
		return 0, false
	end
	self._accumulate_delay_time = self._total_delay_time
	self._accumulate_count = 1
	self._func()
	return time - self._total_delay_time, true
end

function LoopTimer:Update(frame_time)
	if self._accumulate_delay_time < self._total_delay_time then
		self._accumulate_delay_time = self._accumulate_delay_time + (frame_time)
		if self._accumulate_delay_time < self._total_delay_time then
			return
		end
		self._accumulate_delay_time = self._total_delay_time
	end
	if self._accumulate_count >= 1 then
		return
	end
	self._accumulate_count = 1
	self._func()
end

